ethpm/utils/backend.py

```diff
@@ -0,0 +1,38 @@
+from typing import Iterable, Type
+
+from ethpm.backends.base import BaseURIBackend
+from ethpm.backends.ipfs import (
+    DummyIPFSBackend,
+    InfuraIPFSBackend,
+    IPFSGatewayBackend,
+    LocalIPFSBackend,
+    get_ipfs_backend_class,
+)
+from ethpm.backends.registry import RegistryURIBackend
+from ethpm.exceptions import CannotHandleURI
+
+URI_BACKENDS = [
+    IPFSGatewayBackend,
+    InfuraIPFSBackend,
+    DummyIPFSBackend,
+    LocalIPFSBackend,
+    RegistryURIBackend,
+]
+
+
+def get_backends_for_uri(uri: str) -> Iterable[Type[BaseURIBackend]]:
+    default_ipfs = get_ipfs_backend_class()
+    good_backends = [
+        backend
+        for backend in URI_BACKENDS
+        # type ignored because of conflict with instantiating BaseURIBackend
+        if backend().can_handle_uri(uri)  # type: ignore
+    ]
+    if default_ipfs in good_backends:
+        # Move default to front of the list of good backends.
+        good_backends.insert(0, good_backends.pop(good_backends.index(default_ipfs)))
```

Comment:
Not sure if it's better, but here's how I would write this.

```python

def get_backend_for_uri(uri):
    backends = _get_backend_for_uri(uri)
    if not backends:
        raise CannotHandleURI(...)
    return backends

@to_tuple
def _get_backends_for_uri(uri):
    # special case the default IPFS backend to the first slot.
    default_ipfs = get_ipfs_backend_class()
    if default_ipfs in URI_BACKENDS and default_ipfs().can_handle_uri(uri):
        yield default_ipfs
    for backend_class in URI_BACKENDS:
        if backend_class is default_ipfs:
            continue
        elif backend_class().can_handle_uri(uri):
            yield backend_class
```

I *think* the logic of the function is at least *as* easy to understand in this format and it gets rid of the mutation.